File: ros2_project/src/2-1TopicCommunication/one-way_topic_communication_cpp/src/SayHello.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <chrono>
#include <rclcpp/node.hpp>
#include <rclcpp/publisher.hpp>
#include <rclcpp/timer.hpp>

class SayHello : public rclcpp::Node{
public:
    // 构造函数
    SayHello(std::string name):rclcpp::Node(name){
        // 创建发布者
        say_hello = this->create_publisher<std_msgs::msg::String>("/HelloWorld",10);

        // 创建定时器，用于定时发布信息
        say_hello_timer = this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&SayHello::PubHello,this));
    }

private:
    /*
        函数
    */
    void PubHello(){
        // 创建消息
        std_msgs::msg::String msg;
        msg.data = "你好世界";      // 我们可以发现std_msgs::msg::String有一个属性data，而data刚好是C++的std::string类型

        // 发布者发布消息
        say_hello->publish(msg);

        // DEBUG，输出一些东西到终端，让我们知道这个PubHello函数确实被执行了
        RCLCPP_INFO(this->get_logger(),"发布者发布了：%s",msg.data.c_str());    // c_str()表示转换字符串为c语言风格，这是ros2中输出日志的函数需要的
    }

    /*
        声明变量
    */
    // 声明一个发布者指针（其实我们理解为发布者就可以了）
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr say_hello;
    
    // 声明一个定时器指针（其实我们理解为定时器就可以了）
    // 定时器用于根据一定的频率来执行某一个回调函数，我们通过这个回调函数来使Publisher发布消息，就实现发布者按某个频率一直发布消息
    rclcpp::TimerBase::SharedPtr say_hello_timer;

};


int main(int argc, char ** argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SayHello>("PubHello");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_project/src/2-1TopicCommunication/one_way_topic_communication_py/one_way_topic_communication_py/SayHello.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class PubHello(Node):
    # 构造函数，在每个类对象初始化时都会执行
    def __init__(self,name):
        super().__init__(name)

        # 创建一个发布者（Python中不需要声明）
        self.pub_hello = self.create_publisher(String,"/HelloWorld",10)

        # 创建一个定时器，用于定时给发布者发布消息（Python中不需要声明）
        self.timer = self.create_timer(0.5,self.timer_callback)
    
    # 定时器的回调函数
    def timer_callback(self):
        # 构建消息并发布
        msg = String()
        msg.data = "你好世界"
        self.pub_hello.publish(msg)

        # DEBUG，显示当前函数确实在运行，发布者确实在发布
        self.get_logger().info(f"发布者发布消息: {msg.data}")

def main(args=None):
    rclpy.init(args=args)       # 初始化rclpy
    node = PubHello("SayHello") # 新建一个节点
    rclpy.spin(node)            # 保持节点运行，检测是否收到退出指令(Ctrl+C)
    rclpy.shutdown()            # 当接收到退出指令时，退出spin函数，执行shutdown函数来关闭rclpy

